libs/aio/odc/aio/__init__.py

```diff
@@ -292,6 +294,78 @@ def default_guide(url, depth, base):
     return step
 
 
+def s3_find_uri(uri: str, skip_check: bool):
+    """Build generator from supplied S3 URI pattern
+    Arguments:
+        uri {str} -- S3 URI to create generator for
+        skip_check {bool} -- Validity check flag
+    Raises:
+        e: ValueError if parsing URI fails
+    Returns:
+        [type] -- [description]
+    Yields:
+        [type] -- [description]
+    """
+    def do_file_query(qq, pred):
+        for d in s3.dir_dir(qq.base, qq.depth):
+            _, _files = s3.list_dir(d).result()
+            for f in _files:
+                if pred(f):
+                    yield f
+
+    def do_file_query2(qq):
+        fname = qq.file
+
+        stream = s3.dir_dir(qq.base, qq.depth)
+
+        if skip_check:
+            yield from (SimpleNamespace(url=d+fname) for d in stream)
+            return
+
+        stream = (s3.head_object(d+fname) for d in stream)
+
+        for (f, _), _ in future_results(stream, 32):
+            if f is not None:
+                yield f
+
+    def do_dir_query(qq):
+        return (SimpleNamespace(url=url) for url in s3.dir_dir(qq.base, qq.depth))
+
+    try:
+        qq = parse_query(uri)
+    except ValueError as e:
+        logging.error(f"S3 Find Error {ve}")
+        raise e
+
+    s3 = S3Fetcher()
+
+    glob_or_file = qq.glob or qq.file
+
+    if qq.depth is None and glob_or_file is None:
+        stream = s3.find(qq.base)
+    elif qq.depth is None or qq.depth < 0:
+        if qq.glob:
+            stream = s3.find(qq.base, glob=qq.glob)
+        elif qq.file:
+            postfix = '/'+qq.file
+            stream = s3.find(qq.base, pred=lambda o: o.url.endswith(postfix))
+    else:
+        # fixed depth query
+        if qq.glob is not None:
+            pred = norm_predicate(glob=qq.glob)
+            stream = do_file_query(qq, pred)
+        elif qq.file is not None:
+            stream = do_file_query2(qq)
+        else:
+            stream = do_dir_query(qq)
+
+    try:
+        for i, o in enumerate(stream):
```

Comment:
remove enumerate if you don't use it, in fact what is the point of this loop, you can just return iterator like this:
```python
return (o.url for o in stream)
```

In fact why do we need to turn into strings? Why not return a richer object, what if you want to check timestamp or size, turning that to a string is trivial afterwards

Reply:
Actually I'm not sure, what else is in this there except for `.url`?